File: digest.cpp
// read digest_*.txt and generate digest_*.bin

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>

#include "digest.h"

using std::vector;

// str is a string of integers, separated by spaces, ending with newline.
// parse it into a vector
//
inline void parse_ints(const char* str, vector<int> &vec) {
    char *p;
    vec.clear();
    while (1) {
        int i = (int)strtol(str, &p, 10);
        if (p == str) break;
        vec.push_back(i);
        if (!p) break;
        str = p;
    }
}

// str is a string of integers, separated by spaces, ending with newline
// Set the corresponding bits in b
//
inline void set_bits(const char* str, BITMAP& b, int max_id) {
    vector<int> vec;
    parse_ints(str, vec);
    for (int i: vec) {
        if (i>max_id) {
            fprintf(stderr, "bad ID %d\n", i);
            exit(1);
        }
        b.set(i);
    }
}

// str is a string of integers, separated by spaces, ending with newline
// Fill in up to MAX_COMBOS in the given int array
//
void set_inst_combos(const char* str, int* ids) {
    vector<int> vec;
    parse_ints(str, vec);
    for (int i=0; i<MAX_COMBOS; i++) {
        if (i < vec.size()) {
            ids[i] = vec[i];
        } else {
            ids[i] = 0;
        }
    }
}

#define BUF_SIZE    4096

void do_works() {
    FILE* fin = fopen("data/digest_work.txt", "r");
    FILE* fout = fopen("data/digest_work.bin", "w");
    char buf[BUF_SIZE];
    WORK work;
    int n, id;
    char *p, c;
    while (1) {
        p = fgets(buf, BUF_SIZE, fin);
        if (!p) break;
        strncpy(work.title, buf, sizeof(work.title));
        work.title[sizeof(work.title)-1] = 0;

        p = fgets(buf, BUF_SIZE, fin);
        work.work_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &wtb = (BITMAP&)work.work_type;
        wtb.clear(WORK_TYPE_NWORDS);
        set_bits(buf, wtb, WORK_TYPE_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        work.period_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &ib = (BITMAP&)work.inst;
        ib.clear(INSTRUMENT_NWORDS);
        set_bits(buf, ib, INSTRUMENT_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        set_inst_combos(buf, work.inst_combos);

        p = fgets(buf, BUF_SIZE, fin);
        work.comp_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &nb = (BITMAP&)work.comp_nationality;
        nb.clear(NATIONALITY_NWORDS);
        set_bits(buf, nb, NATIONALITY_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        work.comp_sex = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        work.year_of_first_publication = atoi(buf);

        //printf("wrote %s\n", work.title);
        fwrite(&work, sizeof(work), 1, fout);
    }
    fclose(fout);
    fclose(fin);
}

void do_scores() {
    FILE* fin = fopen("data/digest_score.txt", "r");
    FILE* fout = fopen("data/digest_score.bin", "w");
    char buf[BUF_SIZE];
    SCORE score;
    int n, id;
    char *p, c;
    while (1) {
        p = fgets(buf, BUF_SIZE, fin);
        if (!p) break;
        strncpy(score.title, buf, sizeof(score.title));
        score.title[sizeof(score.title)-1] = 0;

        p = fgets(buf, BUF_SIZE, fin);
        score.score_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        score.work_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &wtb = (BITMAP&)score.work_type;
        wtb.clear(WORK_TYPE_NWORDS);
        set_bits(buf, wtb, WORK_TYPE_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        score.period_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &ib = (BITMAP&)score.inst;
        ib.clear(INSTRUMENT_NWORDS);
        set_bits(buf, ib, INSTRUMENT_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        set_inst_combos(buf, score.inst_combos);

        p = fgets(buf, BUF_SIZE, fin);
        score.comp_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &nb = (BITMAP&)score.comp_nationality;
        nb.clear(NATIONALITY_NWORDS);
        set_bits(buf, nb, NATIONALITY_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        score.comp_sex = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        score.license_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        score.is_arrangement = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        score.publisher_id = atoi(buf);

        fwrite(&score, sizeof(score), 1, fout);
    }
    fclose(fout);
    fclose(fin);
}

void do_recs() {
    FILE* fin = fopen("data/digest_rec.txt", "r");
    FILE* fout = fopen("data/digest_rec.bin", "w");
    char buf[BUF_SIZE];
    REC rec;
    int n, id;
    char *p, c;
    while (1) {
        p = fgets(buf, BUF_SIZE, fin);
        if (!p) break;
        strncpy(rec.title, buf, sizeof(rec.title));
        rec.title[sizeof(rec.title)-1] = 0;

        p = fgets(buf, BUF_SIZE, fin);
        rec.rec_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        rec.work_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &wtb = (BITMAP&)rec.work_type;
        wtb.clear(WORK_TYPE_NWORDS);
        set_bits(buf, wtb, WORK_TYPE_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        rec.period_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &ib = (BITMAP&)rec.inst;
        ib.clear(INSTRUMENT_NWORDS);
        set_bits(buf, ib, INSTRUMENT_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        set_inst_combos(buf, rec.inst_combos);

        p = fgets(buf, BUF_SIZE, fin);
        rec.comp_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        BITMAP &nb = (BITMAP&)rec.comp_nationality;
        nb.clear(NATIONALITY_NWORDS);
        set_bits(buf, nb, NATIONALITY_MAX_ID);

        p = fgets(buf, BUF_SIZE, fin);
        rec.comp_sex = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        rec.license_id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        rec.is_arrangement = atoi(buf);

        fwrite(&rec, sizeof(rec), 1, fout);
    }
    fclose(fout);
    fclose(fin);
}

void do_inst_combos() {
    FILE* fin = fopen("data/digest_ic.txt", "r");
    FILE* fout = fopen("data/digest_ic.bin", "w");
    char buf[BUF_SIZE];
    INST_COMBO ic;
    vector<int> vec;
    char *p;
    while (1) {
        memset((void*)&ic, 0, sizeof(ic));
        p = fgets(buf, BUF_SIZE, fin);
        if (!p) break;
        ic.id = atoi(buf);

        p = fgets(buf, BUF_SIZE, fin);
        parse_ints(buf, vec);
        int i=0;
        for (int j: vec) {
            ic.inst_id[i++] = j;
        }

        p = fgets(buf, BUF_SIZE, fin);
        parse_ints(buf, vec);
        i = 0;
        for (int j: vec) {
            ic.count[i++] = j;
        }
        fwrite(&ic, sizeof(ic), 1, fout);
    }
    fclose(fout);
    fclose(fin);
}

int main(int, char**) {
    printf("doing works\n");
    do_works();
    printf("doing scores\n");
    do_scores();
    printf("doing recordings\n");
    do_recs();
    printf("doing inst combos\n");
    do_inst_combos();
}

File: digest.php
#! /usr/bin/env php

<?php

// write flat-file digests of works, scores, recordings, and inst combos
// (digest_*.txt).
// These are then read by digest.cpp and converted to binary files

require_once('imslp_db.inc');
require_once('ser.inc');

// expand a work type list to include descendants
//
function expand_work_type_list($wt_ids) {
    $out = $wt_ids;
    static $work_types = false;
    if (!$work_types) {
        $work_types = work_types();
        foreach ($work_types as $wt) {
            $wt->descendant_ids = json_decode($wt->descendant_ids);
        }
    }

    foreach ($wt_ids as $wt_id) {
        $wt = $work_types[$wt_id];
        if ($wt->descendant_ids) {
            $out = array_unique(array_merge($out, $wt->descendant_ids));
        }
    }
    return $out;
}

function sex_str_to_int($str) {
    if ($str == 'male') return 1;
    if ($str == 'female') return 2;
    return 0;
}

function write_composer_info($f, $work) {
    fprintf($f, "$work->composer_id\n");
    $comp = DB_person::lookup_id($work->composer_id);
    if ($comp->nationality_ids) {
        $nats = json_decode($comp->nationality_ids);
        $nats = array_map('strval', $nats);
        $nats = implode(' ', $nats);
    } else {
        $nats = '';
    }
    fprintf($f, "$nats\n");

    $sex = sex_str_to_int($comp->sex);
    fprintf($f, "$sex\n");
}

function write_work_types($f, $work) {
    if ($work->work_type_ids) {
        $wt_ids = json_decode($work->work_type_ids);
        //$wt_ids = expand_work_type_list($wt_ids);
        $x = implode(' ', $wt_ids);
        fprintf($f, "$x\n");
    } else {
        fprintf($f, "\n");
    }
}

// write instrument IDs, then combo IDs
//
function write_instrument_info($f, $cids) {
    static $inst_combos=false;
    if (!$inst_combos) {
        $inst_combos = get_inst_combos();
    }
    $inst_ids = [];
    foreach ($cids as $id) {
        $ic = $inst_combos[$id];
        $x = json_decode($ic->instruments);
        $inst_ids = array_unique(array_merge($inst_ids, $x->id));
    }
    fprintf($f, implode(' ', array_map('strval', $inst_ids))."\n");

    $x = implode(' ', $cids);
    fprintf($f, "$x\n");
}

function do_works() {
    $f = fopen('data/digest_work.txt', 'w');
    $works = DB_work::enum('', '');
    foreach ($works as $work) {
        fprintf($f, strtolower($work->title)."\n");
        fprintf($f, "$work->id\n");
        write_work_types($f, $work);
        fprintf($f, "$work->period_id\n");
        if ($work->instrument_combo_ids) {
            $cids = json_decode($work->instrument_combo_ids);
            write_instrument_info($f, $cids);
        } else {
            fprintf($f, "\n\n");
        }
        write_composer_info($f, $work);
        fprintf($f, sprintf("%d\n", $work->year_pub));
    }
    fclose($f);
}

function do_score($f, $score) {
    $work = DB_work::lookup_id($score->work_id);
    fprintf($f, strtolower($work->title)."\n");
    fprintf($f, "$score->id\n");
    fprintf($f, "$work->id\n");
    write_work_types($f, $work);
    fprintf($f, "$work->period_id\n");
    if ($score->instrument_combo_ids) {
        $cids = json_decode($score->instrument_combo_ids);
        write_instrument_info($f, $cids);
    } else if ($work->instrument_combo_ids) {
        $cids = json_decode($work->instrument_combo_ids);
        write_instrument_info($f, $cids);
    } else {
        fprintf($f, "\n\n");
    }
    write_composer_info($f, $work);
    fprintf($f, "$score->copyright_id\n");
    if ($score->hier1 == 'Arrangements and Transcriptions') {
        fprintf($f, "1\n");
    } else {
        fprintf($f, "0\n");
    }
    fprintf($f, "$score->publisher_id\n");
}

// do scores 10000 at a time to avoid running out of memory
//
function do_scores() {
    $f = fopen('data/digest_score.txt', 'w');
    $min_id = 0;
    while (1) {
        echo "min_id $min_id\n";
        $scores = DB_score_file_set::enum("id>$min_id", 'limit 10000');
        if (!$scores) break;
        foreach ($scores as $score) {
            do_score($f, $score);
            if ($score->id > $min_id) $min_id = $score->id;
        }
    }
    fclose($f);
}

function do_recs() {
    $f = fopen('data/digest_rec.txt', 'w');
    $recs = DB_audio_file_set::enum('', '');
    foreach ($recs as $rec) {
        $work = DB_work::lookup_id($rec->work_id);
        fprintf($f, strtolower($work->title)."\n");
        fprintf($f, "$rec->id\n");
        fprintf($f, "$work->id\n");
        write_work_types($f, $work);
        fprintf($f, "$work->period_id\n");
        if ($rec->instrument_combo_id) {
            write_instrument_info($f, [$rec->instrument_combo_id]);
        } else if ($work->instrument_combo_ids) {
            $cids = json_decode($work->instrument_combo_ids);
            write_instrument_info($f, $cids);
        } else {
            fprintf($f, "\n\n");
        }
        write_composer_info($f, $work);
        fprintf($f, "$rec->copyright_id\n");
        if ($rec->hier1 == 'Arrangements and Transcriptions') {
            fprintf($f, "1\n");
        } else {
            fprintf($f, "0\n");
        }
    }
    fclose($f);
}

function do_inst_combos() {
    $f = fopen('data/digest_ic.txt', 'w');
    $ics = DB_instrument_combo::enum();
    $max_len = 0;
    foreach ($ics as $ic) {
        $insts = json_decode($ic->instruments);
        array_multisort($insts->id, $insts->count);
        fwrite($f, "$ic->id\n");
        $x = implode(' ', array_map('strval', $insts->id));
        fwrite($f, "$x\n");
        $x = implode(' ', array_map('strval', $insts->count));
        fwrite($f, "$x\n");

        $n = count($insts->id);
        if ($n > $max_len) {
            $max_len = $n;
            $max_id = $ic->id;
        }
    }
    printf("max IC length $max_len; ID $max_id\n");
    fclose($f);
}

echo "doing works\n";
do_works();
echo "doing scores\n";
do_scores();
echo "doing recordings\n";
do_recs();
echo "doing instrument combos\n";
do_inst_combos();

?>
